core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala

```diff
@@ -72,6 +72,18 @@ private[spark] class SortShuffleWriter[K, V, C](
       val partitionLengths = sorter.writePartitionedFile(blockId, tmp)
       shuffleBlockResolver.writeIndexFileAndCommit(dep.shuffleId, mapId, partitionLengths, tmp)
       mapStatus = MapStatus(blockManager.shuffleServerId, partitionLengths)
+      partitionLengths.foreach(writeMetrics.incBlockSizeDistribution(_))
+      if (mapStatus.isInstanceOf[HighlyCompressedMapStatus]) {
+        writeMetrics.setAverageBlockSize(
+          mapStatus.asInstanceOf[HighlyCompressedMapStatus].getAvgSize());
+        (0 until partitionLengths.length).foreach {
+          case i =>
+            if (partitionLengths(i) < mapStatus.getSizeForBlock(i)) {
+              writeMetrics.incUnderestimatedBlocksNum()
+              writeMetrics.incUnderestimatedBlocksSize(partitionLengths(i))
```

Comment:
another metric that may be nice to capture here is *maximum* underestimate -- `(0 until partitionLengths).map { i => partitionLengths(i) - mapStatus.getSizeForBlock(i) }.max`.  In fact, that alone might be enough to discover cases where the reduce side will OOM because of this underestimate.